### PR TITLE
Add troubleshooting section for documentation with widgets

### LINKED ISSUE #519: Widget, requirejs - troublesome again
Jupyter widget provides a nice [cookiecutter template](https://github.com/jupyter-widgets/widget-ts-cookiecutter) that provide a template for generating documentation. The template preloads `jupyter-sphinx` and `nbsphinx`. But if you drop `jupyter-sphinx`, the widgets are not renderer (i did not found a workaround). And if you keep it the widgets are rendered twice (this can be fixed xref #378).

`jupyter-sphinx` is needed because when activating the two extensions (and applying #378), the JS scripts order is (using https://osscar-docs.readthedocs.io/en/latest/widgets/periodictable.html):

image

The trouble is the JS ordering (again :sob: ):
1. Libraries from `jupyter-sphinx`
2. Local bundle with the widget `../_static/embed-periodictable.js`
3. Libraries from `nbsphinx` (only _requirejs_ appears as jupyter widget was drop using #378 fix)

If `jupyter-sphinx` is removed, requireJS is loaded after the local widget bundle. Therefore those widgets are not loaded (error: _define is undefined_).

For now the best configuration I found is to use `jupyter-sphinx` just to ensure requireJS to be loaded before the local widget bundle. And I set `nbsphinx_requirejs_path` and `nbsphinx_widgets_path` to `""` to ensure they are not loaded twice.

I know that requireJS has been moved to a later stage following #409. So I don't rely know what could be a good solution except documenting it. Or maybe you know a better hook than `builder-inited` to insert the local widget bundle.

### PR DESCRIPTION
**Pull request by fcollonval**
Fixes #519

### CODE REVIEW

File: doc/code-cells.ipynb [hunk 1]
```diff
@@ -678,6 +678,40 @@
     "</div>"
    ]
   },
+  {
+   "source": [
+    "#### Troubleshooting\n",
+    "\n",
+    "To obtain more information is widgets are not displayed as expected, you will need to look at the error message in the web browser console.\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:685]: is -> if

File: doc/code-cells.ipynb [hunk 2]
```diff
@@ -678,6 +678,40 @@
...
+    "\n",
+    "> To figure out how to open the web browser console, you may look at the web browser documentation:  \n",
+    "> Chrome: https://developers.google.com/web/tools/chrome-devtools/open#console  \n",
+    "> Firefox: https://developer.mozilla.org/en-US/docs/Tools/Web_Console#Opening_the_Web_Console\n",
+    "\n",
+    "The error is most probably linked to the Javascript files not being loaded or loaded in the wrong order within the html file. To analyze the error, open the HTML file with a text editor and look at the `head` section of the page. That section should contains\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:691]: html -> HTML

Comment (mgeier) [inline doc/code-cells.ipynb:691]: opening the HTML with a text editor might be more complicated than viewing the page source in the browser?

Comment (mgeier) [inline doc/code-cells.ipynb:691]: contains -> contain

File: doc/code-cells.ipynb [hunk 3]
```diff
@@ -678,6 +678,40 @@
...
+    "at least those libraries:\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:692]: Here you say "at least", but below you say that only the first two are mandatory.

**Review by mgeier — COMMENTED** [review]
Thanks a lot for this PR!

I just a few minor comments ...

File: doc/code-cells.ipynb (hunk 2 above)
Comment (fcollonval) [reply doc/code-cells.ipynb:691]: Yep this may indeed be easier. I'll change the sentence.

File: doc/code-cells.ipynb (hunk 3 above)
Comment (fcollonval) [reply doc/code-cells.ipynb:692]: I'll improve the wording.

**Comment by fcollonval** [comment]
Thanks for the review @mgeier

I hope I got your comments right.

File: doc/code-cells.ipynb [hunk 4]
```diff
@@ -678,6 +678,40 @@
...
+   "source": [
+    "#### Troubleshooting\n",
+    "\n",
+    "To obtain more information if widgets are not displayed as expected, you will need to look at the error message in the web browser console.\n",
+    "\n",
+    "> To figure out how to open the web browser console, you may look at the web browser documentation:  \n",
+    "> Chrome: https://developers.google.com/web/tools/chrome-devtools/open#console  \n",
+    "> Firefox: https://developer.mozilla.org/en-US/docs/Tools/Web_Console#Opening_the_Web_Console\n",
+    "\n",
+    "The error is most probably linked to the Javascript files not being loaded or loaded in the wrong order within the HTML file. To analyze the error, you can inspect the HTML file within the web browser (e.g.: right-click on the page and select *View Page Source*) and look at the `head` section of the page. That section should contain\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:691]: Javascript -> JavaScript

File: doc/code-cells.ipynb [hunk 5]
```diff
@@ -678,6 +678,40 @@
...
+    "some Javascript libraries. Those relevant for widgets are:\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:692]: same

File: doc/code-cells.ipynb [hunk 6]
```diff
@@ -678,6 +678,40 @@
...
+    "\n",
+    "```html\n",
+    "<!-- require.js is a mandatory dependency for jupyter-widgets -->\n",
+    "<script crossorigin=\"anonymous\" integrity=\"sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=\" src=\"https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js\"></script>\n",
+    "<!-- jupyter-widgets Javascript -->\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:697]: again

File: doc/code-cells.ipynb [hunk 7]
```diff
@@ -678,6 +678,40 @@
...
+    "<script type=\"text/javascript\" src=\"https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js\"></script>\n",
+    "<!-- Javascript containing custom Jupyter widgets -->\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:699]: and again

File: doc/code-cells.ipynb [hunk 8]
```diff
@@ -678,6 +678,40 @@
...
+    "<script src=\"../_static/embed-widgets.js\"></script>\n",
+    "```\n",
+    "\n",
+    "The two first elements are mandatory. The third one is required only if you designed your own widgets but did not published them on npm.js.\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:703]: published -> publish

**Review by mgeier — COMMENTED** [review]
Thanks for the update! Here are a few more nits:

**Comment by fcollonval** [comment]
Thanks for the English corrections. I made the changes.

FYI: I don't know if you are aware of the [_suggestion_ option](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) in GitHub review:

image

Basically it allows to suggest an code change that will render as diff. Those suggestion can then be batched into a single commit.

It definitely increases productivity as your review comment is directly linked to a code change

File: doc/code-cells.ipynb [hunk 9]
```diff
@@ -678,6 +678,40 @@
...
+   "source": [
+    "#### Troubleshooting\n",
+    "\n",
+    "To obtain more information if widgets are not displayed as expected, you will need to look at the error message in the web browser console.\n",
+    "\n",
+    "> To figure out how to open the web browser console, you may look at the web browser documentation:  \n",
+    "> Chrome: https://developers.google.com/web/tools/chrome-devtools/open#console  \n",
+    "> Firefox: https://developer.mozilla.org/en-US/docs/Tools/Web_Console#Opening_the_Web_Console\n",
+    "\n",
+    "The error is most probably linked to the JavaScript files not being loaded or loaded in the wrong order within the HTML file. To analyze the error, you can inspect the HTML file within the web browser (e.g.: right-click on the page and select *View Page Source*) and look at the `head` section of the page. That section should contain\n",
```
Comment (mgeier) [inline doc/code-cells.ipynb:691]:
```suggestion
    "The error is most probably linked to the JavaScript files not being loaded or loaded in the wrong order within the HTML file. To analyze the error, you can inspect the HTML file within the web browser (e.g.: right-click on the page and select *View Page Source*) and look at the `<head>` section of the page. That section should contain\n",
```

**Review by mgeier — COMMENTED** [review]
Thanks for the suggestion about suggestions!

I have seen this used in other project, but I had never tried it myself.

So no I had to search really hard to find something to try this out:

**Pull request merged**